paasta_tools/deployd/master.py

```diff
@@ -0,0 +1,172 @@
+    def __init__(self):
+        PaastaThread.__init__(self)
+        PaastaThread.daemon = True
+        self.started = False
+        self.bounce_q = PaastaQueue("BounceQueue")
+        self.inbox_q = PaastaQueue("InboxQueue")
+        self.control = PaastaQueue("ControlQueue")
+        self.inbox = Inbox(self.inbox_q, self.bounce_q)
+        self.config = load_system_paasta_config()
+
+    def run(self):
+        self.log.info("paasta-deployd starting up...")
+        with ZookeeperPool() as zk:
+            self.log.info("Waiting to become leader")
+            self.election = PaastaLeaderElection(zk, "/paasta-deployd-leader", socket.getfqdn(), control=self.control)
+            self.is_leader = False
+            self.election.run(self.startup)
+
+    def bounce(self, service_instance):
+        self.inbox_q.put(service_instance)
+
+    def startup(self):
+        self.is_leader = True
+        self.log.debug("This node is elected as leader {}".format(socket.getfqdn()))
+        self.inbox.start()
+        self.log.info("Starting all watcher threads")
+        self.start_watchers()
+        self.log.info("All watchers started, now adding all services for initial bounce")
+        self.add_all_services()
+        self.log.info("Starting worker threads")
+        self.start_workers()
+        self.started = True
+        self.main_loop()
+
+    def main_loop(self):
+        while True:
+            try:
+                message = self.control.get(block=False)
+            except Empty:
+                message = None
+            if message == "ABORT":
+                break
+            time.sleep(0.1)
+
+    def stop(self):
+        self.control.put("ABORT")
+
+    def start_workers(self):
+        for i in range(5):
+            worker = PaastaDeployWorker(i, self.inbox_q, self.bounce_q)
+            worker.start()
+
+    def add_all_services(self):
+        instances = get_services_for_cluster(cluster=self.config.get_cluster(),
+                                             instance_type='marathon',
+                                             soa_dir=DEFAULT_SOA_DIR)
+        instances_to_add = splay_instances(instances=instances,
+                                           splay_minutes=30,
+                                           watcher_name='daemon_start')
+        for service_instance in instances_to_add:
+            self.inbox_q.put(service_instance)
+
+    def start_watchers(self):
+        """ should block until all threads happy"""
+        watcher_classes = [obj[1] for obj in inspect.getmembers(watchers) if inspect.isclass(obj[1]) and
+                           obj[1].__bases__[0] == watchers.PaastaWatcher]
+        self.watchers = [watcher(inbox_q=self.inbox_q,
+                                 cluster=self.config.get_cluster())
+                         for watcher in watcher_classes]
+        self.log.info("Starting the following watchers {}".format(self.watchers))
+        [watcher.start() for watcher in self.watchers]
```

Comment:
do:

```
for watcher in self.watchers:
    watcher.start()
```

since you're really just modelling a side effect, rather than any sort of transformation.